### PR TITLE
Adds all the post API's

### LINKED ISSUE #94: Implementing Add, Delete and Update API's for a post.
**Is your feature request related to a problem? Please describe.**
Implementing the delete and update functionalities for posts on the Donut platform

**Additional context**
I am working on it.

### LINKED ISSUE #101: Implement the GET API's for the post module.
**Is your feature request related to a problem? Please describe.**
Implementing all the GET API's for the post module.

**Additional context**
Working on it.

### LINKED ISSUE #115: [ENHANCEMENT]Writing Unit Tests for the Post API's
**Describe the enhancement you'd like**
Writing Unit Tests for the Post API's. Tests will be done for all the routes in the Post Module.

**Additional context**
Working on this issue

### PR DESCRIPTION
**Pull request by rak-shit**
---
name: Adds all the post API's
about: Implements the backend of post API's
title: 'Adds all the post API's'
labels: 'feature'
assignees: @rak-shit 

---

### **Problem**
Github Issue Number: Fixes #101, Fixes #94 & Fixes #115

### **Type of Change**
[ ] Bug fix
[x] New Feature
[ ] Development of UI/UX prototypes
[ ] Small refactor
[ ] Change in Documentation

### **Checklist**
[x] My code follows the same style as the codebase
[ ] My Code change requires a change in documentation
[ ] I have updated the Readme accordingly
[x] I made PR against **development branch**
[ ] I have run the test cases locally and it's passing.
[x] I have squashed my commits

### CODE REVIEW

File: app/controllers/post.js [hunk 1]
```diff
@@ -26,7 +99,60 @@ module.exports = {
...
+      })
+      // CHECKS IF THE USER HAS ALREADY DOWNVOTED THE COMMENT
+      post.votes.upVotes.users.user.filter(user => {
+        if (JSON.stringify(user) === JSON.stringify(userId)) {
+          post.votes.upVotes.count = post.votes.upVotes.count - 1
+          post.votes.upVotes.users.user.remove(user)
+        }
+      })
+      post.votes.downVotes.count = post.votes.downVotes.count + 1
+      post.votes.downVotes.users.user.push(userId)
```
Comment (devesh-verma) [inline app/controllers/post.js:151]: why are we not increasing the downVotes count here?

File: app/models/Post.js [hunk 2]
```diff
@@ -33,11 +33,10 @@ const PostSchema = new Schema({
         default: 0
       },
       users: {
-        user: {
+        user: [{
```
Comment (devesh-verma) [inline app/models/Post.js:36]: why are you doing this? as in users then inside that user array. cant we just use users and have array of objects inside that?

File: app/models/Post.js [hunk 3]
```diff
@@ -46,11 +45,10 @@ const PostSchema = new Schema({
...
```
Comment (devesh-verma) [inline app/models/Post.js:48]: same here

File: app/controllers/post.js [hunk 4]
```diff
@@ -1,16 +1,89 @@
 const userModel = require('../models/User')
 const bcrypt = require('bcrypt')
 const jwt = require('jsonwebtoken')
+const PostModel = require('../models/Post')
+const HANDLER = require('../utils/response-helper')
+const STATUS = require('../utils/status-codes')
+const helper = require('../utils/uploader')
```
Comment (devesh-verma) [inline app/controllers/post.js:7]: make the constant name as `imgUploadHelper`

File: app/routes/post.js [hunk 5]
```diff
@@ -1,35 +1,60 @@
...
+  userController.getPostById
+)
+
+router.put(
+  '/upvote/:id/:userId',
+  userController.upvote
+)
+
+router.put(
+  '/downvote/:id/:userId',
```
Comment (devesh-verma) [inline app/routes/post.js:56]: this should be just `/downvote/:id`

File: app/routes/post.js [hunk 6]
```diff
@@ -1,35 +1,60 @@
...
+
+// GET TASK BY ID
+router.get(
+  '/:id',
+  auth,
+  userController.getPostById
+)
+
+router.put(
+  '/upvote/:id/:userId',
```
Comment (devesh-verma) [inline app/routes/post.js:51]: this should be just `/upvote/:id`

File: app/controllers/post.js [hunk 7]
```diff
@@ -1,16 +1,89 @@
...
+  },
+  delete: async (req, res, next) => {
+    const { id } = req.params
+    console.log(id)
+    try {
+      const post = await PostModel.findByIdAndRemove(id)
+      if (!post) {
+        return res.status(STATUS.NOT_FOUND).json({ message: 'No post exists' })
       }
+      res.status(STATUS.OK).json({ post: post, message: 'Deleted the post' })
```
Comment (devesh-verma) [inline app/controllers/post.js:30]: you are missing a very imp part here. You who owns a post can only delete his post or if her is an admin then can delete the post/

File: app/models/Post.js (hunk 2 above)
Comment (rak-shit) [reply app/models/Post.js:36]: I did not want to disturb the predefined schema. I was also thinking that having just users and have an array of objects inside that would be better. Will do it that way.

File: app/controllers/post.js (hunk 4 above)
Comment (rak-shit) [reply app/controllers/post.js:7]: Sure :)

File: app/controllers/post.js (hunk 7 above)
Comment (rak-shit) [reply app/controllers/post.js:30]: Oh yes, I will do it.

File: app/routes/post.js (hunk 6 above)
Comment (rak-shit) [reply app/routes/post.js:51]: Then, how do we store the user who is making the upvote? Should I send both userID and the commentID in the same parameter and then split them?

File: app/controllers/post.js (hunk 1 above)
Comment (rak-shit) [reply app/controllers/post.js:151]: The flow goes this way, first we check if the user has already downvoted the comment, if yes we take that as a bad request and it ends here, if not, then we check if the user has already upvoted the comment, if yes we first remove the upvote and then add the downvote count, if not, we directly add the downvote count.

File: app/controllers/post.js [hunk 8]
```diff
@@ -1,16 +1,89 @@
...
+        } if (!posts.length) {
+          return res.status(STATUS.NOT_FOUND).json({ message: 'Item not found' })
+        }
+        return res.status(STATUS.OK).json({ success: true, data: posts })
+      })
+    } catch (error) {
+      HANDLER.handleError(res, error)
+    }
   },
   authenticate: function (req, res, next) {
```
Comment (devesh-verma) [inline app/controllers/post.js:88]: why do we have this function here?

Comment (rak-shit) [reply app/controllers/post.js:88]: That was there earlier, I did not touch it anyway. I will remove it, seems like it is of no use there.

**Comment by rak-shit** [comment]
**Post creation**

Screenshot 2020-04-11 at 5 18 07 PM
Screenshot 2020-04-11 at 5 18 15 PM

**GET all posts**

Screenshot 2020-04-11 at 5 27 14 PM

**GET post by ID**

Screenshot 2020-04-11 at 5 27 34 PM

**Delete post**

Screenshot 2020-04-11 at 5 29 36 PM

**Downvote**

Screenshot 2020-04-11 at 5 31 36 PM

**Upvote**

Screenshot 2020-04-11 at 5 31 49 PM

**Pull request merged**